File: app/src/main/java/com/artifex/mupdfdemo/OutlineActivity.java
package com.artifex.mupdfdemo;

import android.app.Activity;
import android.os.Build;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toolbar;

public class OutlineActivity extends Activity {
	OutlineItem mItems[];

	/** Called when the activity is first created. */
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.outline_list);

		if (Build.VERSION.SDK_INT>=Build.VERSION_CODES.LOLLIPOP) {
			Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
			//toolbar.setLogo(R.drawable.icon);
			setActionBar(toolbar);
			getActionBar().setDisplayHomeAsUpEnabled(true);
			/*toolbar.setNavigationOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					finish();
				}
			});*/
		} else if (Build.VERSION.SDK_INT>=Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
			getActionBar().setDisplayHomeAsUpEnabled(true);
            getActionBar().setDisplayShowHomeEnabled(true);
        } else {
            setTheme(R.style.AppFullscreen);
        }

		final ListView listView=(ListView) findViewById(R.id.list);
		listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
			@Override
			public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
				onListItemClick((ListView) parent, view, position, id);
			}
		});
		mItems = OutlineActivityData.get().items;
		listView.setAdapter(new OutlineAdapter(getLayoutInflater(), mItems));
		// Restore the position within the list from last viewing

		//getListView().setDividerHeight(0);
		//listView.setCacheColorHint(getResources().getColor(android.R.color.transparent));
		setResult(-1);
		listView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
			@Override
			public void onGlobalLayout() {
				listView.getViewTreeObserver().removeGlobalOnLayoutListener(this);
				setSelection(listView);
			}
		});
	}

	private void setSelection(ListView listView) {
		int pos = OutlineActivityData.get().position;
		if (null != getIntent()) {
            int cp = getIntent().getIntExtra("cp", -1);
            if (cp != -1) {
                pos = cp;
            }
        }
		OutlineItem[] items = OutlineActivityData.get().items;
		//System.out.println("pos:"+pos+" items:"+items.length);
		OutlineItem item;
		int idx=0;
		for (int i = items.length - 1; i >= 0; i--) {
            item = items[i];
            //System.out.println("pos page:"+item.page);
            if (item.page <= pos) {
                //pos = item.page;
                idx=i;
                break;
            }
        }

		//System.out.println("pos:"+pos+" idx:"+idx);
		if (idx < listView.getAdapter().getCount()) {
            listView.setSelection(idx);
        }
	}

	protected void onListItemClick(ListView l, View v, int position, long id) {
		OutlineActivityData.get().position = l.getFirstVisiblePosition();
		setResult(mItems[position].page);
		finish();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int itemId=item.getItemId();
		if (itemId==android.R.id.home) {
			finish();
		}

		return super.onOptionsItemSelected(item);
	}
}

File: app/src/main/java/com/artifex/mupdfdemo/OutlineActivity.kt
package com.artifex.mupdfdemo

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewTreeObserver
import android.widget.AdapterView
import android.widget.ListView
import android.widget.Toolbar

import cn.archko.pdf.R

class OutlineActivity() : Activity() {

    var mItems: Array<OutlineItem>? = null

    /** Called when the activity is first created.  */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.outline_list)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val toolbar = findViewById(R.id.toolbar) as Toolbar
            setActionBar(toolbar)
            actionBar!!.setDisplayHomeAsUpEnabled(true)
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            actionBar!!.setDisplayHomeAsUpEnabled(true)
            actionBar!!.setDisplayShowHomeEnabled(true)
        } else {
            setTheme(R.style.AppFullscreen)
        }

        val listView = findViewById(R.id.list) as ListView
        listView.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id -> onListItemClick(parent as ListView, view, position, id) }

        mItems = OutlineActivityData.get().items
        listView.adapter = OutlineAdapter(layoutInflater, mItems)

        setResult(-1)
        listView.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                listView.viewTreeObserver.removeGlobalOnLayoutListener(this)
                setSelection(listView)
            }
        })
    }

    private fun setSelection(listView: ListView) {
        var pos = OutlineActivityData.get().position
        if (null != intent) {
            val cp = intent.getIntExtra("cp", -1)
            if (cp != -1) {
                pos = cp
            }
        }
        val items = OutlineActivityData.get().items
        var item: OutlineItem
        var idx = 0
        for (i in items.indices.reversed()) {
            item = items[i]
            if (item.page <= pos) {
                idx = i
                break
            }
        }

        if (idx < listView.adapter.count) {
            listView.setSelection(idx)
        }
    }

    protected fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        OutlineActivityData.get().position = l.firstVisiblePosition
        setResult(mItems!![position].page)
        finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        if (itemId == android.R.id.home) {
            finish()
        }

        return super.onOptionsItemSelected(item)
    }
}
